bk-nodes/bk_db.py

```diff
@@ -0,0 +1,241 @@
+
+import MySQLdb
+import config
+import dateutil.parser
+
+
+table_fields = {}
+table_fields_index ={}
+
+
+class BeekeeperDB():
+    def __init__ ( self , retries=60) :
+
+        if not config.mysql_host:
+            raise Exception("mysql_host is not defined")
+        
+        count = 0
+        while True:
+            try:
+                self.db=MySQLdb.connect(host=config.mysql_host,user=config.mysql_user,
+                  passwd=config.mysql_password,db=config.mysql_db)
+            except Exception as e: # pragma: no cover
+                if count > retries:
+                    raise
+                print(f'Could not connnect to database, error={e}, retry in 2 seconds', file=sys.stderr)
+                time.sleep(2)
+                count += 1
+                continue
+            break
+
+        self.cur=self.db.cursor()
+        return
+
+
+    def nodes_log_add(self, logData):
+        #node_id, table_name, operation, field_name, new_value, source, effective_time=None
+        fields = '`node_id`, `table_name`, `operation`, `field_name`, `new_value`, `source`, `effective_time`'
+        values_s = '%s, %s, %s, %s, %s, %s, %s'
+        #values = [node_id, table_name, operation, field_name, new_value, source]
+        
+        smallest_timestamp = None 
+        values = []
+        for log in logData:
+            effective_time =  dateutil.parser.parse(log["effective_time"])
+            
+            if (not smallest_timestamp) or (effective_time < smallest_timestamp):
+                smallest_timestamp = effective_time
+
+            value_tuple =(log["node_id"], log["table_name"], log["operation"], log["field_name"], log["new_value"], log["source"], effective_time)
+            values.append(value_tuple)
+
+
+
+
+        self.cur.execute("LOCK TABLES `nodes_log` WRITE")
+
+        stmt = f'INSERT INTO `nodes_log` ( {fields} ) VALUES ({values_s})'
+        #debug_stmt = stmt
+        for tup in values:
+            debug_stmt = stmt
+            for i in tup:
+                debug_stmt = debug_stmt.replace("%s", f'"{i}"', 1)
+            print(f'debug_stmt: {debug_stmt}', flush=True)
+
+        #print(f'debug_stmt: {debug_stmt}', file=sys.stderr)
+        
+        self.cur.executemany(stmt, values )
+        #self.cur.execute("UNLOCK TABLES")
```

Comment:
I think you need this (but moved to the end of the function).

you lock the table above.  I know that the above call to `replay_log()` will perform an `UNLOCK` if everything goes well.  But what about the failure condition where `replay_log()` throws an exception.

It seems that no matter what happens you should perform an UNLOCK at the end of this function right before the return.  I *assume* that executing an `UNLOCK` on an already unlocked table is a NOP.

Reply:
An exception will destroy the mysql connection, when this happens all tables are unlocked. But you are right, it seems I forgot the unlock. I added it.